### PR TITLE
add support for chunk strategys, currently support: randomAccess, Str…

### PR DESCRIPTION
**Pull request by shabtaisharon**
…eam & noAllocate

### CODE REVIEW

**Comment by shabtaisharon** [comment]
* Ignore the printing, i still need them for the debug 
* Still need to add more tests, currently the default works with all the past tests we had

**Comment by shabtaisharon** [comment]
refactor the DataIntegrity to use all the current strategys

**Comment by shabtaisharon** [comment]
refacor tests to use all strategys more elegantly

File: Ds3/Helpers/Strategys/ChunkStrategys/IChunkStrategy.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+* ****************************************************************************
+*/
+
+using Ds3.Calls;
+using Ds3.Helpers.TransferItemSources;
+using System.Collections.Generic;
+
+namespace Ds3.Helpers.Strategys.ChunkStrategys
+{
+    public interface IChunkStrategy
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/ChunkStrategys/IChunkStrategy.cs]: Please fill out the code comments here to explain what the interface methods do.

File: Ds3/Helpers/Strategys/ReadRandomAccessHelperStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        private IStreamFactory<TItem> _readRandomAccessStreamFactory;
+
+        public ReadRandomAccessHelperStrategy(int retyrAfter)
+        {
+            this._retryAfter = retyrAfter;
+        }
+
+        public IChunkStrategy GetChunkStrategy()
+        {
+            lock (_lock)
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/ReadRandomAccessHelperStrategy.cs]: Why the lock?

**Comment by shabtaisharon** [comment]
some design changes

File: Ds3/Helpers/Jobs/Job.cs [hunk 3]
```diff
@@ -33,18 +34,23 @@ internal abstract class Job<TSelf, TItem> : IBaseJob<TSelf, TItem>
...
-        private readonly ILookup<Blob, Range> _rangesForRequests;
-        private readonly IRangeTranslator<Blob, TItem> _rangeTranslator;
-        private readonly JobItemTracker<TItem> _itemTracker;
+        protected readonly ITransferrer _transferrer;
+        protected readonly ILookup<Blob, Range> _rangesForRequests;
+        protected readonly IRangeTranslator<Blob, TItem> _rangeTranslator;
+        protected readonly JobItemTracker<TItem> _itemTracker;
 
-        private IResourceStore<TItem, Stream> _resourceStore;
+        //protected IResourceStore<TItem, Stream> _resourceStore;
```
Comment (rpmoore) [inline Ds3/Helpers/Jobs/Job.cs]: delete this.

File: Ds3/Helpers/Strategys/ChunkStrategys/IChunkStrategy.cs (hunk 1 above)
Comment (rpmoore) [inline Ds3/Helpers/Strategys/ChunkStrategys/IChunkStrategy.cs]: You filled out one of the code comments.  Please do the same for the other 2 methods.

File: Ds3/Helpers/Strategys/ChunkStrategys/WriteNoAllocateChunkStrategy.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+*/
+
+using Ds3.Calls;
+using Ds3.Helpers.TransferItemSources;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Ds3.Helpers.Strategys.ChunkStrategys
+{
+    internal class WriteNoAllocateChunkStrategy : IChunkStrategy
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/ChunkStrategys/WriteNoAllocateChunkStrategy.cs]: This probably should not be made internal in the event that a user wants to create their own IHelperStrategy with this ChunkStrategy in it.

Comment (rpmoore) [inline Ds3/Helpers/Strategys/ChunkStrategys/WriteNoAllocateChunkStrategy.cs]: Also, please add some code comments to explain this strategy.

File: Ds3/Helpers/Strategys/ChunkStrategys/WriteRandomAccessChunkStrategy.cs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+using Ds3.Helpers.TransferItemSources;
+using Ds3.Models;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace Ds3.Helpers.Strategys.ChunkStrategys
+{
+    internal class WriteRandomAccessChunkStrategy : IChunkStrategy
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/ChunkStrategys/WriteRandomAccessChunkStrategy.cs]: Same comment about the visibility of this class and comments

File: Ds3/Helpers/Strategys/ChunkStrategys/WriteStreamChunkStrategy.cs [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+using Ds3.Helpers.TransferItemSources;
+using Ds3.Models;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace Ds3.Helpers.Strategys.ChunkStrategys
+{
+    internal class WriteStreamChunkStrategy : IChunkStrategy
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/ChunkStrategys/WriteStreamChunkStrategy.cs]: Same comment about the visibility of this class and comments

File: Ds3/Helpers/Strategys/StreamFactory/IStreamFactory.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        /// </summary>
+        /// <param name="createStreamForTransferItem"></param>
+        /// <param name="rangeTranslator"></param>
+        /// <param name="blob"></param>
+        /// <param name="length"></param>
+        /// <returns></returns>
+        Stream CreateStream(Func<TItem, Stream> createStreamForTransferItem, IRangeTranslator<Blob, TItem> rangeTranslator, Blob blob, long length);
+
+        /// <summary>
+        ///
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/StreamFactory/IStreamFactory.cs]: Please fill out the comment for the 2 remaining methods in this interface.

File: Ds3/Helpers/Strategys/StreamFactory/ReadRandomAccessStreamFactory.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+*/
+
+using Ds3.Helpers.RangeTranslators;
+using Ds3.Helpers.Streams;
+using System;
+using System.IO;
+
+namespace Ds3.Helpers.Strategys.StreamFactory
+{
+    internal class ReadRandomAccessStreamFactory<TItem> : IStreamFactory<TItem>
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/StreamFactory/ReadRandomAccessStreamFactory.cs]: Same comment about the visibility of this class and comments

File: Ds3/Helpers/Strategys/StreamFactory/StreamFactoryException.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+*   CONDITIONS OF ANY KIND, either express or implied. See the License for the
+*   specific language governing permissions and limitations under the License.
+* ****************************************************************************
+*/
+
+using System;
+
+namespace Ds3.Helpers.Strategys.StreamFactory
+{
+    internal class StreamFactoryException : Exception
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/StreamFactory/StreamFactoryException.cs]: Should this be made public?

File: Ds3/Helpers/Strategys/StreamFactory/WriteRandomAccessStreamFactory.cs [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+using Ds3.Helpers.RangeTranslators;
+using Ds3.Helpers.Streams;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+
+namespace Ds3.Helpers.Strategys.StreamFactory
+{
+    internal class WriteRandomAccessStreamFactory : IStreamFactory<string>
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/StreamFactory/WriteRandomAccessStreamFactory.cs]: Same comment about the visibility of this class and comments

File: Ds3/Helpers/Strategys/StreamFactory/WriteStreamStreamFactory.cs [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+using Ds3.Helpers.RangeTranslators;
+using Ds3.Helpers.Streams;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+
+namespace Ds3.Helpers.Strategys.StreamFactory
+{
+    internal class WriteStreamStreamFactory : IStreamFactory<string>
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/StreamFactory/WriteStreamStreamFactory.cs]: Same comment about the visibility of this class and comments

File: Ds3/Helpers/Strategys/WriteNoAllocateHelperStrategy.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+*   specific language governing permissions and limitations under the License.
+* ****************************************************************************
+*/
+
+using Ds3.Helpers.Strategys.ChunkStrategys;
+using Ds3.Helpers.Strategys.StreamFactory;
+
+namespace Ds3.Helpers.Strategys
+{
+    public class WriteNoAllocateHelperStrategy : IHelperStrategy<string>
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/WriteNoAllocateHelperStrategy.cs]: Please add a comment explaining what this class does.

File: Ds3/Helpers/Strategys/WriteRandomAccessHelperStrategy.cs [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+*   specific language governing permissions and limitations under the License.
+* ****************************************************************************
+*/
+
+using Ds3.Helpers.Strategys.ChunkStrategys;
+using Ds3.Helpers.Strategys.StreamFactory;
+
+namespace Ds3.Helpers.Strategys
+{
+    public class WriteRandomAccessHelperStrategy : IHelperStrategy<string>
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/WriteRandomAccessHelperStrategy.cs]: Please add a comment explaining what this class does.

File: Ds3/Helpers/Strategys/WriteStreamHelperStrategy.cs [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+*   specific language governing permissions and limitations under the License.
+* ****************************************************************************
+*/
+
+using Ds3.Helpers.Strategys.ChunkStrategys;
+using Ds3.Helpers.Strategys.StreamFactory;
+
+namespace Ds3.Helpers.Strategys
+{
+    public class WriteStreamHelperStrategy : IHelperStrategy<string>
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/WriteStreamHelperStrategy.cs]: Comments

File: Ds3/Helpers/TransferItemSources/ITransferItemSource.cs [hunk 15]
```diff
@@ -17,6 +17,7 @@
 
 namespace Ds3.Helpers.TransferItemSources
 {
+    //TODO delete this class
```
Comment (rpmoore) [inline Ds3/Helpers/TransferItemSources/ITransferItemSource.cs]: So delete the class?

File: Ds3/Helpers/TransferItemSources/ReadTransferItemSource.cs [hunk 16]
```diff
@@ -22,6 +22,7 @@
 
 namespace Ds3.Helpers.TransferItemSources
 {
+    //TODO delete this class
```
Comment (rpmoore) [inline Ds3/Helpers/TransferItemSources/ReadTransferItemSource.cs]: Delete the class

File: Ds3/Helpers/TransferItemSources/WriteTransferItemSource.cs [hunk 17]
```diff
@@ -22,6 +22,7 @@
 
 namespace Ds3.Helpers.TransferItemSources
 {
+    //TODO delete this class
```
Comment (rpmoore) [inline Ds3/Helpers/TransferItemSources/WriteTransferItemSource.cs]: Delete! :D

File: IntegrationTestDS3/Ds3TestUtils.cs [hunk 18]
```diff
@@ -13,27 +13,34 @@
...
 using System.Security.Cryptography;
 
 namespace IntegrationTestDS3
 {
     internal static class Ds3TestUtils
     {
-
         public static IDs3Client CreateClient()
         {
+            //TODO remove me
```
Comment (rpmoore) [inline IntegrationTestDS3/Ds3TestUtils.cs]: Delete....

File: TestDs3/Helpers/TestDs3ClientHelpers.cs [hunk 19]
```diff
@@ -274,6 +275,8 @@ public void BasicWriteTransfer(long? maxBlobSize)
             job.DataTransferred += dataTransfers.Enqueue;
             job.ItemCompleted += itemsCompleted.Enqueue;
 
+            //TODO need to find out why fail with more threads
```
Comment (rpmoore) [inline TestDs3/Helpers/TestDs3ClientHelpers.cs]: Do you have any more information on this?

File: Ds3/Helpers/TransferItemSources/ITransferItemSource.cs (hunk 15 above)
Comment (shabtaisharon) [inline Ds3/Helpers/TransferItemSources/ITransferItemSource.cs]: can't delete it until the unit test is changed

File: Ds3/Helpers/TransferItemSources/ReadTransferItemSource.cs (hunk 16 above)
Comment (shabtaisharon) [inline Ds3/Helpers/TransferItemSources/ReadTransferItemSource.cs]: can't delete it until the unit test is changed

File: Ds3/Helpers/TransferItemSources/WriteTransferItemSource.cs (hunk 17 above)
Comment (shabtaisharon) [inline Ds3/Helpers/TransferItemSources/WriteTransferItemSource.cs]: can't delete it until the unit test is changed

File: IntegrationTestDS3/Ds3TestUtils.cs (hunk 18 above)
Comment (shabtaisharon) [inline IntegrationTestDS3/Ds3TestUtils.cs]: will be removed

File: TestDs3/Helpers/TestDs3ClientHelpers.cs (hunk 19 above)
Comment (shabtaisharon) [inline TestDs3/Helpers/TestDs3ClientHelpers.cs]: not yet

File: Ds3/Helpers/Strategys/ChunkStrategys/WriteStreamChunkStrategy.cs [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+using Ds3.Models;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace Ds3.Helpers.Strategys.ChunkStrategys
+{
+    /// <summary>
+    /// The WriteStreamChunkStrategy will allocate chunk at a time for each stream in the job and return the allocated blobs
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/ChunkStrategys/WriteStreamChunkStrategy.cs]: 'chunks one at a time'

File: Ds3/Helpers/Strategys/StreamFactory/WriteStreamStreamFactory.cs [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+        private readonly Dictionary<string, Stream> _streamStore = new Dictionary<string, Stream>();
+
+        public Stream CreateStream(Func<string, Stream> createStreamForTransferItem, IRangeTranslator<Blob, string> rangeTranslator, Blob blob, long length)
+        {
+            lock (_lock)
+            {
+                Stream stream;
+                if (this._streamStore.TryGetValue(blob.Context, out stream))
+                {
+                    //TODO add a way to check the position of the stream when using a non seeking stream
```
Comment (rpmoore) [inline Ds3/Helpers/Strategys/StreamFactory/WriteStreamStreamFactory.cs]: We discussed this, and it can be removed.

Comment (shabtaisharon) [inline Ds3/Helpers/Strategys/StreamFactory/WriteStreamStreamFactory.cs]: ok, i'll remove the comment

File: TestDs3/Helpers/Strategys/ChunkStrategys/TestReadRandomAccessChunkStrategy.cs [hunk 22]
```diff
@@ -237,70 +237,74 @@ public void TestEnumerateTransfersSetRetryAfter()
...
                 .Returns(factory.Object);
             client
-                .Setup(c => c.GetAvailableJobChunks(AvailableChunks(Stubs.JobId)))
+                .Setup(c => c.GetAvailableJobChunks(AllocateMock.AvailableChunks(Stubs.JobId)))
                 .Returns(GetAvailableJobChunksResponse.RetryAfter(
                     TimeSpan.FromMinutes(0)));
 
-            var transferItemSource = new ReadTransferItemSource(_ => { }, client.Object, 0, initialJobResponse);
-            using (var transfers = transferItemSource.EnumerateAvailableTransfers().GetEnumerator())
+            //var transferItemSource = new ReadTransferItemSource(_ => { }, client.Object, 0, initialJobResponse);
```
Comment (rpmoore) [inline TestDs3/Helpers/Strategys/ChunkStrategys/TestReadRandomAccessChunkStrategy.cs]: delete this.

File: TestDs3/Helpers/Strategys/ChunkStrategys/TestReadRandomAccessChunkStrategy.cs [hunk 23]
```diff
@@ -237,70 +237,74 @@ public void TestEnumerateTransfersSetRetryAfter()
...
                 {
-                    Assert.True(transferItemSource.RetryAfterLeft == 0);
+                    Assert.True(source.RetryAfterLeft == 0);
                     Assert.True(ex.Message.Equals("Reached the limit number of retries request"));
                 }
             }
 
-            transferItemSource = new ReadTransferItemSource(_ => { }, client.Object, 1, initialJobResponse);
-            using (var transfers = transferItemSource.EnumerateAvailableTransfers().GetEnumerator())
+            //transferItemSource = new ReadTransferItemSource(_ => { }, client.Object, 1, initialJobResponse);
```
Comment (rpmoore) [inline TestDs3/Helpers/Strategys/ChunkStrategys/TestReadRandomAccessChunkStrategy.cs]: delete this

File: TestDs3/Helpers/Strategys/ChunkStrategys/TestReadRandomAccessChunkStrategy.cs [hunk 24]
```diff
@@ -237,70 +237,74 @@ public void TestEnumerateTransfersSetRetryAfter()
...
                 {
-                    Assert.True(transferItemSource.RetryAfterLeft == 0);
+                    Assert.True(source.RetryAfterLeft == 0);
                     Assert.True(ex.Message.Equals("Reached the limit number of retries request"));
                 }
             }
 
-            transferItemSource = new ReadTransferItemSource(_ => { }, client.Object, 2, initialJobResponse);
-            using (var transfers = transferItemSource.EnumerateAvailableTransfers().GetEnumerator())
+            //transferItemSource = new ReadTransferItemSource(_ => { }, client.Object, 2, initialJobResponse);
```
Comment (rpmoore) [inline TestDs3/Helpers/Strategys/ChunkStrategys/TestReadRandomAccessChunkStrategy.cs]: ditto

File: TestDs3/Helpers/Strategys/ChunkStrategys/TestReadRandomAccessChunkStrategy.cs [hunk 25]
```diff
@@ -237,70 +237,74 @@ public void TestEnumerateTransfersSetRetryAfter()
...
                 {
-                    Assert.True(transferItemSource.RetryAfterLeft == 0);
+                    Assert.True(source.RetryAfterLeft == 0);
                     Assert.True(ex.Message.Equals("Reached the limit number of retries request"));
                 }
             }
 
-            transferItemSource = new ReadTransferItemSource(_ => { }, client.Object, 100, initialJobResponse);
-            using (var transfers = transferItemSource.EnumerateAvailableTransfers().GetEnumerator())
+            //transferItemSource = new ReadTransferItemSource(_ => { }, client.Object, 100, initialJobResponse);
```
Comment (rpmoore) [inline TestDs3/Helpers/Strategys/ChunkStrategys/TestReadRandomAccessChunkStrategy.cs]: ditto

File: TestDs3/Helpers/Strategys/ChunkStrategys/TestReadRandomAccessChunkStrategy.cs [hunk 26]
```diff
@@ -336,18 +340,25 @@ public void TestEnumerateTransfersResetRetryAfter()
                     return GetAvailableJobChunksResponse.RetryAfter(TimeSpan.FromSeconds(0));
                 });
 
-            var transferItemSource = new ReadTransferItemSource(_ => { retryAfter--; }, client.Object, retryAfter, initialJobResponse);
+            //var transferItemSource = new ReadTransferItemSource(_ => { retryAfter--; }, client.Object, retryAfter, initialJobResponse);
```
Comment (rpmoore) [inline TestDs3/Helpers/Strategys/ChunkStrategys/TestReadRandomAccessChunkStrategy.cs]: ditto again

**Pull request merged**